File: 0721-accounts-merge/0721-accounts-merge.py
class Solution:
    def findUpar(self, x):
        if self.par[x] == x:
            return x
        self.par[x] = self.findUpar(self.par[x])
        return self.par[x]
    def unionByRank(self, x, y):
        ulp_x = self.findUpar(x)
        ulp_y = self.findUpar(y)
        if ulp_x == ulp_y:
            return
        if self.rank[ulp_x] < self.rank[ulp_y]:
            par[ulp_x] = ulp_y
        elif self.rank[ulp_x] > self.rank[ulp_y]:
            par[ulp_y] = ulp_x
        else:
            self.par[ulp_x] = ulp_y
            self.rank[ulp_x] += 1
        
    def accountsMerge(self, accounts: list[list[str]]) -> list[list[str]]:
        n = len(accounts)
        self.par = list(range(n))
        self.rank = [0] * n
        map = {}
        for i in range(n):
            self.par[i] = i
        for i in range(n):
            for j in range(1, len(accounts[i])):
                mail = accounts[i][j]
                if mail not in map:
                    map[mail] = i
                else:
                    self.unionByRank(i, map[mail])
        mergedAccounts = [[] for _ in range(n)]
        for emails, accounts_index in map.items():
            node = self.findUpar(accounts_index)
            mergedAccounts[node].append(emails)
        
        ans = []
        for i in range(n):
            if len(mergedAccounts[i]) == 0:
                continue
            temp = []
            temp.append(accounts[i][0])
            mergedAccounts[i].sort()
            temp.extend(mergedAccounts[i])
            ans.append(temp)

        return ans

File: 0721-accounts-merge/0721-accounts-merge.cpp
class Solution {
public:
    vector<int> parent;
    vector<int> size;

    int findUpar(int node) {
        if (node == parent[node]) {
            return node;
        }
        return parent[node] = findUpar(parent[node]);
    }

    void unionBySize(int x, int y) {
        int ulp_x = findUpar(x);
        int ulp_y = findUpar(y);
        if (ulp_x == ulp_y) {
            return;
        }
        if (size[x] > size[y]) {
            parent[ulp_y] = ulp_x;
            size[x] += size[y];
        } else if (size[x] < size[y]) {
            parent[ulp_x] = ulp_y;
            size[y] += size[x];
        } else {
            parent[ulp_x] = ulp_y;
            size[y] += size[x];
        }
    }
    vector<vector<string>> accountsMerge(vector<vector<string>>& accounts) {
        int n = accounts.size();
        parent.resize(n);
        size.resize(n, 1);
        for(int i=0; i<n; i++){
            parent[i] = i;
        }
        unordered_map<string, int> mp;
        for (int i = 0; i < accounts.size(); i++) {
            for (int j = 1; j < accounts[i].size(); j++) {
                if (mp.find(accounts[i][j]) == mp.end()){
                    mp[accounts[i][j]] = i;
                }
                else {
                    unionBySize(i, mp[accounts[i][j]]);
                }
            }
        }
        vector<vector<string>> mergedAccounts(accounts.size());
        for (auto& it : mp) {
            string mail = it.first;
            int node = findUpar(it.second);
            mergedAccounts[node].push_back(mail);
        }
        vector<vector<string>> ans;
        for (int i = 0; i < accounts.size(); i++) {
            if (mergedAccounts[i].size() == 0)
                continue;
            sort(mergedAccounts[i].begin(), mergedAccounts[i].end());
            vector<string> temp;
            temp.push_back(accounts[i][0]);
            for (auto it : mergedAccounts[i]) {
                temp.push_back(it);
            }
            ans.push_back(temp);
        }
        return ans;
    }
};